core/src/main/java/org/infinispan/conflict/impl/DefaultConflictManager.java

```diff
@@ -249,7 +245,9 @@ public void resolveConflicts(EntryMergePolicy<K, V> mergePolicy) {
       } else {
          localizedTopology = distributionManager.createLocalizedCacheTopology(topology);
       }
-      conflictFuture = CompletableFuture.runAsync(() -> doResolveConflicts(localizedTopology, entryMergePolicy, preferredNodes), resolutionExecutor);
+      conflictFuture = resolutionExecutor.execute(() -> doResolveConflicts(localizedTopology, entryMergePolicy, preferredNodes),
+            localizedTopology.getTopologyId())
+            .toCompletableFuture();
```

Comment:
The important part is that the returned Stage never runs chained methods on a blocking thread, unless the invoking thread is blocking.

Reply:
Also I can't change `conflictFuture` to a stage as we cancel it possibly.

Reply:
> The important part is that the returned Stage never runs chained methods on a blocking thread, unless the invoking thread is blocking.

Fair enough, without it you'd have to call `continueOnNonBlockingThread()`. TBT I was thinking of removing `LimitedExecutor` and removing its usages with `ActionSequencer`, and this new interface interferes with my plans :)

> Also I can't change conflictFuture to a stage as we cancel it possibly.

Cancelling the `CompletableFuture` doesn't do anything to cancel the actual conflict resolution, so I fear cancellation is broken anyway /cc @ryanemerson .

Reply:
> > The important part is that the returned Stage never runs chained methods on a blocking thread, unless the invoking thread is blocking.
> 
> Fair enough, without it you'd have to call `continueOnNonBlockingThread()`.

Yeah the problem I have is that this will be forgotten somewhere. Where as by banning the use of blocking executor and requiring blocking manager, means it won't  be messed up. I am trying to take a hard stance to catching these things before they become problems, by using checkstyle and block hound.

> TBT I was thinking of removing `LimitedExecutor` and removing its usages with `ActionSequencer`, and this new interface interferes with my plans :)

I think we can still do that. I am more than fine removing `BlockingExecutor` as needed. I had ideas of that a while back, but never looked into it.

Reply:
True, it's not public API so it's not that hard to change.

Reply:
I created https://issues.redhat.com/browse/ISPN-11718 for the conflict resolution cancellation.